WalletWasabi/WabiSabi/Client/RoundStateUpdater.cs

```diff
@@ -0,0 +1,142 @@
+			var removedRoundStates = RoundStates.Where(round => !responseRoundStates.ContainsKey(round.Key));
+
+			var roundsToUpdate = updatedRoundStates.Where(updatedRound => RoundStates[updatedRound.Key] != updatedRound.Value)
+				.Union(newRoundStates)
+				.Union(removedRoundStates)
+				.Select(rs => rs.Key)
+				.ToList();
+
+			RoundStates = updatedRoundStates.Union(newRoundStates).ToDictionary(s => s.Key, s => s.Value);
+
+			lock (AwaitersLock)
+			{
+				// Handle round independent tasks.
+				if (Awaiters.TryGetValue(AnyRoundId, out var roundStateAwaiters))
+				{
+					foreach (var roundState in RoundStates.Values)
+					{
+						RemoveCompletedAwaiters(roundStateAwaiters, roundState);
+					}
+				}
+
+				if (roundsToUpdate.Any())
+				{
+					// Handle round dependent tasks.
+					foreach (var roundId in roundsToUpdate)
+					{
+						if (!RoundStates.TryGetValue(roundId, out var roundState))
+						{
+							// The round is missing.
+							var tasks = Awaiters[roundId];
+							foreach (var t in tasks)
+							{
+								t.TaskCompletionSource.TrySetException(new InvalidOperationException($"Round {roundId} is not running anymore."));
+							}
+							Awaiters.Remove(roundId);
+							continue;
+						}
+
+						if (Awaiters.TryGetValue(roundId, out var list))
+						{
+							RemoveCompletedAwaiters(list, roundState);
+						}
+					}
+				}
+			}
+		}
+
+		private static void RemoveCompletedAwaiters(List<RoundStateAwaiter> roundStateAwaiters, RoundState roundState)
+		{
+			foreach (var roundStateAwaiter in roundStateAwaiters.Where(roundStateAwaiter => roundStateAwaiter.Predicate(roundState)).ToArray())
+			{
+				// The predicate was fulfilled.
+				var task = roundStateAwaiter.TaskCompletionSource;
+				task.TrySetResult(roundState);
+				roundStateAwaiters.Remove(roundStateAwaiter);
+			}
+		}
+
+		public Task<RoundState> CreateRoundAwaiter(uint256 roundId, Predicate<RoundState> predicate, CancellationToken cancellationToken)
+		{
+			TaskCompletionSource<RoundState> tcs = new(TaskCreationOptions.RunContinuationsAsynchronously);
+			List<RoundStateAwaiter>? roundStateAwaiters = null;
+			RoundStateAwaiter? roundStateAwaiter = null;
+
+			lock (AwaitersLock)
+			{
+				if (!Awaiters.ContainsKey(roundId))
+				{
+					Awaiters.Add(roundId, new List<RoundStateAwaiter>());
+				}
+				roundStateAwaiters = Awaiters[roundId];
+
+				roundStateAwaiter = new RoundStateAwaiter(tcs, predicate);
+				roundStateAwaiters.Add(roundStateAwaiter);
+			}
+
+			cancellationToken.Register(() =>
+			{
+				tcs.TrySetCanceled();
+				lock (AwaitersLock)
+				{
+					roundStateAwaiters.Remove(roundStateAwaiter);
+				}
+			});
+
+			return tcs.Task;
+		}
+
+		public Task<RoundState> CreateRoundAwaiter(Predicate<RoundState> predicate, CancellationToken cancellationToken)
+		{
+			// Zero denotes that the predicate should run for any round.
+			return CreateRoundAwaiter(AnyRoundId, predicate, cancellationToken);
+		}
```

Comment:
What is this needed for? I can imagine two use cases for the API, but neither needs this:

1. get parameters about all rounds in order to decide which rounds to possibly join
2. get details/updates about a specific round for coinjoin client

or maybe it's for 1 when you care about finding the first round that meets some criteria (e.g. feerate?)

Reply:
It is used here: https://github.com/molnard/WalletWasabi/blob/25068eb077f2defc188f8485144db2cd1db9bdb5/WalletWasabi/WabiSabi/Client/CoinJoinClient.cs#L55

Reply:
so for that use case i imagined needing all the rounds valid rounds to choose between them, i thought maybe it would make sense if that's the only use case but i changed my mind and returning just one is better for now, can resolve